### PR TITLE
Fix assert in the Date prototype

### PR DESCRIPTION
**Pull request by kkristof**
Fix for issue #348

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp [hunk 1]
```diff
@@ -825,6 +825,33 @@ ecma_date_timezone_offset (ecma_number_t time) /**< time value */
...
+ * See also:
+ *          ECMA-262 v5, 15.9.1.1
+ *
+ * Used by:
+ *         - The ecma_date_make_day Date helper routine.
+ *
+ * @return  clamped year to 1970-285,616 and 1970+285,616 range
+ */
+ecma_number_t
+ecma_date_clamp_year (ecma_number_t year)
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: @kkristof, could you, please, add comment to the argument?

**Comment by kkristof** [comment]
Hi, I have updated the branch.

**Comment by ruben-ayrapetyan** [comment]
@kkristof, thank you for the update.

**Comment by ruben-ayrapetyan** [comment]
@kkristof, there is a test case that seems to be not processed according to ECMA specification:
```
d = new Date(8.64e+15);
print (d);
d = new Date(8.64e+15 + 1);
print (d);
```

Output:
```
275760-09-13T00:00:00.000
275760-09-13T00:00:00.001
```

Correct output:
```
275760-09-13T00:00:00.000
Invalid Date
```

**Comment by ruben-ayrapetyan** [comment]
Sorry, I've unintentionally pressed "Closed and comment" instead of "Comment", so pull request is now closed and reopened.

**Comment by ruben-ayrapetyan** [comment]
I think that may be we should add `TimeClip` call somewhere for the test case from https://github.com/Samsung/jerryscript/pull/459#issuecomment-124176046 to be processed correctly.

May be this would also help to solve the assertion, described in the pull request.

**Comment by ruben-ayrapetyan** [comment]
> May be this would also help to solve the assertion, described in the pull request.

No, seems that the issue is not related to the pull request.

**Comment by ruben-ayrapetyan** [comment]
Could you, please, describe in comments why current implementation of `ecma_date_make_day` correctly performs 15.9.1.12, step 7?

**Comment by kkristof** [comment]
Hmm, it seems the implementation of the ecma_date_make_day doesn't match with the standard. Thank you for pointing this out. I'm going to fix it.

**Comment by kkristof** [comment]
Hi, I have updated the branch.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp [hunk 2]
```diff
@@ -741,17 +741,20 @@ ecma_date_make_day (ecma_number_t year, /**< year value */
...
-  JERRY_ASSERT (ecma_date_year_from_time (time) == ym);
-
-  while (ecma_date_month_from_time (time) < mn)
+  if (ecma_date_year_from_time (time) == ym)
   {
-    time += ECMA_DATE_MS_PER_DAY;
-  }
+    while (ecma_date_month_from_time (time) < mn)
+    {
+      time += ECMA_DATE_MS_PER_DAY;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: Could we remove the the loop by calculating necessary difference and adding it at once, so improving performance?

If not, we could:
- add 28 days (minimal number of days in a month) at first step, as initial day number is 1, so reducing number of iterations to at most 3;
- or use binary search for calculating `time` value.

**Comment by ruben-ayrapetyan** [comment]
@kkristof, thank you for update!

Could you, please, add comment with explanation of algorithm used in `ecma_date_make_day`, i.e. description of why the performed steps are equivalent to behaviour, specified in 15.9.1.12 step 7 (https://github.com/Samsung/jerryscript/pull/459#issuecomment-124178575)?

**Comment by kkristof** [comment]
Hi, I have updated the branch.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp [hunk 3]
```diff
@@ -733,25 +734,58 @@ ecma_date_make_day (ecma_number_t year, /**< year value */
...
+  /* 6. */
   ecma_number_t mn = (ecma_number_t) fmod (m, 12);
   mn = (mn < 0) ? 12 + mn : mn;
-  ecma_number_t time = ecma_date_time_from_year (ym);
 
-  JERRY_ASSERT (ecma_date_year_from_time (time) == ym);
+  /* 7. We need to find the time where: YearFromTime(time) == ym
+   *    and MonthFromTime(time) == mn and DateFromTime(time) == 1.
+   *    If this is not possible return NaN.
+   */
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: @kkristof, the comment is requirement for algorithm below (I think description of the requirement could be replaced with `/* 7. */` comment in the case).

However, there is no explanation of why the algorithm below does really find value, described in the comment / specification, i.e. there is no description of why the steps performed below are equivalent to behaviour, specified in 15.9.1.12 step 7.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp [hunk 4]
```diff
@@ -733,25 +734,58 @@ ecma_date_make_day (ecma_number_t year, /**< year value */
...
+      }
+      delta = floor (delta / 2);
+    }
+
+    if (ecma_date_month_from_time (time + step) < mn)
+    {
+      time += step * ECMA_DATE_MS_PER_DAY;
+    }
+
+    while (ecma_date_month_from_time (time) < mn)
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: Shouldn't this loop be removed?

Comment (kkristof) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: No, because the binary search has to use integer days so the approximation is given by the binary search needs some fine tuning here.

**Comment by kkristof** [comment]
Hi, I have rebased and updated the branch with comments.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp [hunk 5]
```diff
@@ -733,25 +734,66 @@ ecma_date_make_day (ecma_number_t year, /**< year value */
...
+  ecma_number_t step = floor ((ecma_number_t) (365) / 2);
+  ecma_number_t delta = step;
+
+  /**
+   * The algorithm below searches the following date: ym-mn-1
+   * To find this time it starts from the beggining of the year (ym)
+   * then increase it by ECMA_DATE_MS_PER_DAY until it reaches the
+   * right date.
+   */
+  if (ecma_date_year_from_time (time) == ym)
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: @galpeter could you please revise this block for performance. I suppose that algorithm can be significantly improved.

Comment (kkristof) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: Hi, I have made a quick test with every date in a year.
The binary search always takes 9 iterations and in the worst case it needs 4 iterations in the day by day phase while in the best case it skips it. The whole test took 0.458s.
The old algorithm is far worse because it needs 335 (december) iterations at the worst case but 0 in the best case (january). The test took 0.574s.
We could use lookup table for the months to get the days but I think we won't get any measurable performance gain in normal usage with this function.

Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: @egavrin , I think it's fine for now, it's faster than the original one. If we want to further improve it then creating a lookup table would be a good idea imho, but that should be a separate PR as this is for fixing the assert. Maybe we can raise an issue for it.

**Comment by galpeter** [comment]
lgtm

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp (hunk 5 above)
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: For example, `floor ((ecma_number_t) (365) / 2)` constant could be calculated more efficiently. Also, maybe there are some other places in the patch that can be improved. By the way, why do we need floating point for [[Date]] values?

Comment (kkristof) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: The `Date` needs 52 significant bit to represent the date as the standard describes it. Probably it would be better to use `int64_t` especially on those target where the `ecma_numer_t` is single precision. I think this refactoring should be in different PR with other optimizations as well, but I'm not sure how much performance gain we would get through optimizing the `Date` object.

Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp]: @kkristof, I agree with you that switching to `uint64_t` should not be performed in this PR.

**Comment by kkristof** [comment]
Hi, I have rebased and updated the branch.

**Comment by egavrin** [comment]
Ok, lets push it. But Date should be optimized a bit: according to `perf` ~20% of test262 execution time is Date-related staff. Please, raise issue for performance optimization of Date.

**Comment by kkristof** [comment]
I have raised #560 for this. I haven't thought that it can cause this much impact.

**Comment by egavrin** [comment]
`make push`
